File: compiler/bench/xorsum.fs
\ expected: 0
\ XOR checksum - compute xor checksum over many iterations

create test-data 256 allot
: init-data ( -- )
  256 0 do i test-data i + c! loop ;

: xorsum ( addr len -- checksum )
  0 swap 0 do
    over i + c@ xor
  loop nip ;

: main
  init-data
  0
  500000 0 do
    test-data 256 xorsum
    xor
  loop
  . cr ;

File: compiler/bench/xorsum.c
// expected: 0
// XOR checksum - compute xor checksum over many iterations
#include <stdio.h>
#include <stdint.h>

unsigned char test_data[256];

uint32_t xorsum(unsigned char *data, int len) {
    uint32_t result = 0;
    for (int i = 0; i < len; i++) {
        result ^= data[i];
    }
    return result;
}

int main(void) {
    for (int i = 0; i < 256; i++) test_data[i] = i;
    uint32_t result = 0;
    for (int i = 0; i < 500000; i++) {
        result ^= xorsum(test_data, 256);
    }
    printf("%u\n", result);
    return 0;
}
